LFS-680: When clicking (+) on a subject chart to add a form, provide list of forms that take this type of subject and its children only

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-680)

This work also includes the ProgenyProcessor, which allows for a new `.json` selector (`.progeny`) which recursively finds the children of a given node (via recursively looking for nodes whose `parent` is the given node). 

**To test**: When viewing the patient chart for a Tumour, trying to create a new form from that page will no longer show Questionnaires like `Demographics`, which cannot be used for a Tumour.

![image](https://user-images.githubusercontent.com/4656440/105753700-19670300-5f17-11eb-8bdf-cb6d957659e3.png)

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Add a search bar to the NewFormDialog subject search
Add the ProgenyServlet, which allows you to search for the progeny of a
SubjectType

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Hookup the new form dialog to our new progeny servlet

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Code cleanup

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Fix the code commenting

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Switch from a servlet to a json processor.

acrowthe commented on modules/utils/src/main/java/ca/sickkids/ccm/lfs/serialize/internal/ProgenyProcessor.java:
```diff
@@ -0,0 +1,181 @@
+            Map<String, Boolean> seenTypes = new HashMap<String, Boolean>();
+            JsonObject progeny = getAllProgeny(node.getPath(), seenTypes, true);
+            json.add("@progeny", progeny.getJsonArray("@progeny"));
+            LOGGER.warn(progeny.toString());
```
Is this log still needed? Doesn't seem to be output in testing

OrdiNeu replied:
Fixed in `d732263` :heavy_check_mark:

acrowthe reviewed (commented):
Looks good and worked as expected

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Remove debugging code

sashaandjic commented:
Should we be showing the "+" button on the Tumor region charts? 
![image](https://user-images.githubusercontent.com/23243244/105776845-5216d480-5f37-11eb-8891-cd6191dab380.png)

sashaandjic commented:
Problem found, steps to reproduce:
- create 2 patients (e.g. `ABC` and `123`)
- for each patient create at least one tumor (e.g. `ABC/lung`, `ABC/liver`, `123/liver`,` 123/brain`)
- open any tumor chart (e.g. `ABC/liver`)
- click on the `+` button
- select Radio form
- expected behavior: Radiotherapy form should be created for the tumor shown in the chart
- current behaviour: dialog pops up offering selection of other patients and their tumors
 
![image](https://user-images.githubusercontent.com/23243244/105778145-bc307900-5f39-11eb-841c-1791f0ada7a5.png)

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Fix a bug that occurs when trying to get the progeny of a node with no
progeny

marta- commented on modules/utils/src/main/java/ca/sickkids/ccm/lfs/serialize/internal/ProgenyProcessor.java:
```diff
@@ -140,10 +140,11 @@ public JsonObject getAllProgeny(String subjectNodePath, Map<String, Boolean> see
-            builder.add("@progeny", arrayBuilder.build());
+        // Copy over this node's info, if any. An empty root node will need an @progeny child.
+        if (hasResult || isRoot) {
+        builder.add("@progeny", arrayBuilder.build());
```
Wrong indentation. Also, the string "@progeny" appears 3 times in this file. It should be a constant.

OrdiNeu replied:
Fixed in `8793a69` :heavy_check_mark:

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

Code refactor

sashaandjic commented:
the bug is still present
![image](https://user-images.githubusercontent.com/23243244/106211620-137f5500-6197-11eb-8546-4a4f99cb5b76.png)

sashaandjic commented:
The problem with tumor region's for selection is now resolved - only Somatic Variants are shown in the selection dialog
![image](https://user-images.githubusercontent.com/23243244/106211849-62c58580-6197-11eb-8035-a7cc450a5100.png)

OrdiNeu commented:
Closed due to being handled in #459
